### ISSUE TITLE
Question: pnet and using [#packet]

### ISSUE THREAD

**Issue by janrunePaessler** [libpnet/libpnet#380]
Newbie question about how to use [#packet]...
I've been banging my head on this for a few days, so I figured I'd ask, so sorry if it should be obvious...
I tend to jump in on the deep end to figure things out, so here we go...

I'm following the docs in "libpnet/docs/using_packet.md" trying to make wire-level packets using the macro's. 
Are those guidelines current ? They were written over 2 years ago, but I could not find any more recent usages. (Though I found some interesting samples in the Redox-OS source.)
```toml
[build-dependencies]
glob = "0.3.0"
regex = "1.0.*"
syntex = "0.42.0"
## pnet_macro_0.22.0 requires syntex_0.42.0
pnet_macros = "0.22.0" 

#[dependencies]
pnet = "0.22.0"
```
## pnet_macro_0.22.0 requires syntex_0.42.0
Figuring out the dependency chain on this one took a while...

myproto.rs.in
```rust
extern crate pnet;
extern crate pnet_macros_support;
use pnet_macros_support::types::{u16be, u32be};
use pnet::pnet_packet::{Packet, PacketData, MutablePacket};
...
#[packet]
pub struct ServerGreeting {
	pub Unused      : [u8; 12],
	pub Modes       : u32be,
	pub Challenge   : [u8; 16],
	pub Salt        : [u8; 16],
	pub Count       : u32be,
	pub MBZ         : [u8; 16],
}
...
#[packet]
pub struct SenderAuthPacket {
	pub SeqNumber     : u32be,
	#[length = "12"]
	pub MBZ1          : Vec<u8>,
//	pub Time          : NTPTSFormat,
//	pub Time   		  : [u32be; 2],
    pub TimeSec	  	  : u32be,
    pub TimeFrac	  : u32be,
	pub ErrorEstimate : u16be,
	#[length = "6"]
	pub MBZ2          : Vec<u8>,
	#[length = "16"]
	pub HMAC          : Vec<u8>,
	#[payload]
    pub Padding       : Vec<u8>
}
```
Looking at the generated code I figured out some of the issues
Issues:
 use pnet_macros_support::types::{u16be, u32be};
does not seem to work, I get undefiend on usages of u16be, u32be
So I had to define:
  pub type u16be = u16;
  pub type u32be = u32;
to make those go away.

Crate issue:
```
error[E0463]: can't find crate for `pnet`
```
2418 | impl <'a> ::pnet_macros_support::packet::Packet for
     |             ^^^^^^^^^^^^^^^^^^^ maybe a missing `extern crate pnet_macros_support;`?
error[E0433]: failed to resolve: maybe a missing `extern crate pnet_macros_support;`?
warning: unused import: `::pnet_macros_support::packet::PacketData`
   --> target\debug\build\myproto-f735887fca4612dd\out/twamp.rs:631:13
631 |         use ::pnet_macros_support::packet::PacketData;

warning: unused import: `::pnet_macros_support::packet::MutPacketData`
   --> target\debug\build\myproto-f735887fca4612dd\out/twamp.rs:710:17
710 |             use ::pnet_macros_support::packet::MutPacketData;

Now for the usage questions...
error: #[packet]'s must contain a payload
   --> src\packet\myproto.rs.in:113:1
If the #[packet] macro is only for packets with variable payload, what do we use for fixed packets ?

Nested Structs:
I have NTP timestamps in my packets:
pub struct NTPTSFormat {
    pub seconds: u32be,
    pub fraction: u32be
}

Is there a way to use it c-Style :
#[packet]
pub struct ServerStart {
	pub MBZ1        : [u8; 15],
	pub Accept      : u8,
	pub ServerIV    : [u8; 16],
	pub StartTime   : NTPTSFormat,
	pub SequenceNo : u16be
}

BitFields:
I have some bitFields I need to decode, like C-Style
struct MSG_Error_Estimate {
	byte s : 1,
	byte z : 1,
	byte scale : 6,
	byte multiplier,
}
What is the best practices for decoding those ?

I would appreciate any insight you may have to share...
 Thanks
  JR

**Comment by JuxhinDB** [comment]
Hey @janrunePaessler -- thanks for reaching out, sorry to hear you're experiencing issues with the library. A couple things before I help dive deep into this:

- What exactly are you looking to achieve (high-level)? 
- Would you be able to format the bottom-half of the ticket as it's hard to follow
- Though it does not use `pnet_macros` you can maybe look at [this](https://github.com/JuxhinDB/synner) repo to get some ideas depending on your use-case

Also, are you trying to port an existing project? If so you could share it here to maybe get a better idea as to what you are trying to achieve.

Thank you

**Comment by janrunePaessler** [comment]
Hi JuxhinDB.

Thanks for getting back to me.

-   A fully functional TWAMP implementation(rfc5357). 
  (like https://github.com/nokia/twampy/blob/master/twampy.py)
-   Implement using tokyo reactors.
-   I will publish the functional implementation here and probably in Redox...

I was trying to use the #packet macro to utilize the provided facilities for dealing with 
variable length payloads and "auto-magical" byte-swap.
Is the #packet macro is not really used/maintained ? I can't find many usages of it, except in the old doc example.
If so, then these issues go away.
The question then is, is there a way to unpack/byte-swap using library code ?
Any good Rust protobuf examples ?

 Thanks
   JR

**Comment by JuxhinDB** [comment]
@janrunePaessler -- unfortunately I'm not too familiar with `#packet_macro`  so it could be the case that it's simply not maintained. 

Regarding protobuf, there appears to be a [stable compiler](https://github.com/stepancheg/rust-protobuf) that you can use along with [serde](https://github.com/serde-rs/serde) -- not sure about examples.

Lastly, byte-swapping magic tends to happen with `pnet_macros` from what I could find so I don't think there is any other library code that handles this. You might try out std lib's [to/from_be](https://doc.rust-lang.org/std/primitive.i32.html#method.from_be) for example but I am not sure if this satisfies your use-case.

**Comment by janrunePaessler** [comment]
Hi JuxhinDB.

Thanks for getting back to me.
Re: #packet_macro , no worries, I can do it "Old-School"...
 Any restrictions of which crates to include in pnetlib ?

I'll research the protobuf/serde options as well.

The to/from_be should work, but my goal is to use tokyo, so I'll find some good examples on Encode/Decode packets... There are some good examples in https://github.com/rusticata, but they only decode...
Would have been nice if they also encoded...

Thanks for your assistance.
  JR

**Comment by JuxhinDB** [comment]
Glad to help @janrunePaessler. Just to add, should you come up with some interesting example feel free to share them here or to twitter directly (@juxhindb) as I'd love to read more about it.
